core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -48,7 +49,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
           .getInputMetricsForReadMethod(blockResult.readMethod)
         existingMetrics.incBytesRead(blockResult.bytes)
 
-        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        val buf = blockResult.data.toArray
```

Comment:
yes, that's right.  Using this approach, and avoiding OOMs requires some more changes -- as you suggest, using `unrollSafely`, or perhaps making a variant of `blockManager.putIterator` to give an iterator back or something like that.  But if the block is only cached to disk anyway, then you end up jumping through some hoops for nothing, so it gets tricky.

All of these complications really have me in favor of "idea 1", for its simplicity, which is why I was avoiding getting into some of the nitty gritty here.  But if we really go for "idea 2", there are a lot more details like this to figure out, I think.